### ISSUE TITLE
Wifi support for ESP32/ESP8266? 

### ISSUE THREAD

**Issue by Technerder** [tinygo-org/tinygo#2024]
I recently stumbled upon this project and was thrilled to see that some basic features such as I2C and SPI were supported on the ESP32/ESP8266. Are there any plans to support Wifi on the chips and if so when can we expect to see a bare/basic implementation in the main code base?

**Comment by gedw99** [comment]
i am also hoping for this... :)

**Comment by aykevl** [comment]
I would absolutely love to see WiFi/BT support on these chips! However, it's going to be difficult because it doesn't have a lot of documentation. I've ordered an ESP32C3 which I want to try to add WiFi support to. Specifically the ESP32C3 because it uses the well documented RISC-V architecture instead of the sparsely documented (if at all) Xtensa architecture in the original ESP32. So my idea is that it should be a lot easier for that reason and when that works, I can use that knowledge for the original ESP32.

I can't give you a time for when it will be finished, like I can't give you a time for when _anything_ will be finished. Sorry.

**Comment by gedw99** [comment]
New hardware news 

https://www.reddit.com/r/homeautomation/comments/owzpou/espressif_to_release_new_esp32h2_microcontroller/?utm_source=share&utm_medium=ios_app&utm_name=iossmf

**Comment by Dorbmon** [comment]
I hope too...

**Comment by scroot** [comment]
i am also hoping for this...

**Comment by wlwatkins** [comment]
any update on this?

**Comment by aykevl** [comment]
Yes, please follow #2138 and https://github.com/tinygo-org/drivers/pull/320.

**Comment by ninanomenon** [comment]
thank you @aykevl for your work! Are there any updates on this?

**Comment by ngandalf** [comment]
is there any news about wifi support on esp8266 / esp32 ?

**Comment by ghost** [comment]
You would be better off using ESP-AT firmware on your ESP and control it with something well supported like RP2040 (yes, i realize thats 2 chips but at least you can have something that works now).
ESP8266/ESP32/ESP32-S2 are no longer actively developed or supported by Espressif, they now only recommend using ESP32-C2/C3 chips.

This is what i ended up doing.

**Comment by ericmigi** [comment]
@conotto I would love to chat about how you solved this - any chance you could DM me on twitter.com/ericmigi or send me an email (username@gmail.com)?

**Comment by CrazyPandar** [comment]
I found an ESP wifi driver, maybe this helps. 

https://github.com/esp-rs/esp-wireless-drivers-3rdparty

**Comment by aykevl** [comment]
@CrazyPandar yes that's what we should use (see https://github.com/tinygo-org/drivers/pull/320#issuecomment-1284080185).

**Comment by DatanoiseTV** [comment]
In rust, this was implemented: https://github.com/esp-rs/esp-wifi
Maybe this could be adopted to TinyGo?

**Comment by aykevl** [comment]
@DatanoiseTV yes, see my comment above (which links to https://github.com/tinygo-org/drivers/pull/320#issuecomment-1284080185). TL;DR: we should indeed use https://github.com/esp-rs/esp-wireless-drivers-3rdparty. But that's a lot of work and nobody has stepped up to do it yet.

**Comment by pcting** [comment]
what would be the process of integrating the [esp-wireless-drivers-3rdparty](https://github.com/esp-rs/esp-wireless-drivers-3rdparty) code? i could give it a try, but i don't have much embedded experience. are we talking about linking in the `.a` files and generating go code to wrap functions in the c header files using `cgo`?

**Comment by aykevl** [comment]
@pcting please follow this PR instead: https://github.com/tinygo-org/drivers/pull/650
Once the ESP32-C3 is implemented, we could take a look at supporting the ESP32/ESP8266 but those are going to be a lot more work because they also need Xtensa interrupt support. I spent a few days trying to get Xtensa interrupt support working at some point, but gave up on it.

I recommend following the ESP32-C3 work.

**Comment by mamoit** [comment]
There seems to have been a significant development on this topic:
https://media.ccc.de/v/38c3-liberating-wi-fi-on-the-esp32

**Comment by aykevl** [comment]
@mamoit true, but it doesn't help us now. Many parts are still missing, so it's not something we can use right now.

(I was there in the room when it was presented, it was an interesting talk!)

**Comment by dector** [comment]
Is it possible to donate to fund work on this issue somehow? It would be awesome.

Thanks for your awesome work btw.

**Comment by deadprogram** [comment]
@dector yes, please, and thank you!

https://opencollective.com/tinygo

**Comment by dector** [comment]
> [@dector](https://github.com/dector) yes, please, and thank you!
> 
> https://opencollective.com/tinygo

Awesome. I will use it.

But as I understand - that's generic contribution to the project.
I was wondering if you are generally considering possibility to accept funds for focused work of specific feature (if enough people will want to contribute for such common goal). I know that not all OSS contributors want do it, so just asking - no pressure.

**Comment by JunBys** [comment]
Excuse me, is there any progress on this problem?

**Comment by aykevl** [comment]
> Yes, please follow [#2138](https://github.com/tinygo-org/tinygo/pull/2138) and [tinygo-org/drivers#320](https://github.com/tinygo-org/drivers/pull/320).

@JunBys please follow the above PRs. I'll post an update when I get to work on it again.

**Comment by niemeyer** [comment]
> [@dector](https://github.com/dector) yes, please, and thank you!
> 
> https://opencollective.com/tinygo

Just became a monthly contributor as well. It's not much at all, but at least it's one more voice towards radio support in cheap and widely available boards, as that tends to be the deal breaker for many use cases.  ESP32 S/C and RP2040/2350 are probably some of the most interesting candidates in that regard, at the moment.

Thanks for your long term work in TinyGo.

**Comment by officer47p** [comment]
Hey @aykevl , I have some free bandwidth and wanted to spend it on this feature.
If you wanted to onboard a developer that you just hired, how would you guide him?
To be honest, just looking at the PRs is not enough and I would need some help in order to help back.

I do want to write go on my esp boards and connect them to internet, just how I do it using python, jealousy is my fuel:)

Looking forward to a reply.

**Comment by aykevl** [comment]
Hi @officer47p thank you for your interest! You can perhaps message me in the Gophers Slack? Or message me in some other way: https://aykevl.nl/about
Then I can get you started with wifi support, at least as far as I got last time.

**Comment by dimajolkin** [comment]
I'm leaving a link to MR here for history.
https://github.com/tinygo-org/espradio/pull/1
I have successfully launched WiFi on the esp32c3. I think it is already clear to others how to connect and what is needed.
